src/contrasts.jl

```diff
@@ -466,49 +474,79 @@ expressed as a hypothesis matrix, but it is not obvious just from looking at the
 contrasts matrix.  For this reason `HypothesisCoding` is preferred for
 specifying custom contrast coding schemes over `ContrastsCoding`.
 
+Optional arguments `levels` and `labels` give the ordering of mapping between
+levels and columns of the hypothesis matrix, and the labels for the generated
+contrasts, respetively.
 """
-mutable struct HypothesisCoding <: AbstractContrasts
-    hypotheses::Matrix
-    contrasts::Matrix
-    base::Nothing
-    levels::Union{Vector,Nothing}
-
-    function HypothesisCoding(hypotheses, base, levels)
+mutable struct HypothesisCoding{T<:AbstractMatrix, S<:AbstractMatrix} <: AbstractContrasts
+    hypotheses::T
+    contrasts::S
+    levels::Union{AbstractVector,Nothing}
```

Comment:
Why parameterize on `S` and `T` but not on the type of this field?

Reply:
Great question.  It didn't seem like it would buy anything performance wise since it's going through the ContrastsMatrix anyway before it's used, and I was starting to get tired of so many type parameters...

Reply:
OK, I just wanted to check there was a (good) reason.